### PR TITLE
Fix servicebinding controller reconciler bug

### LINKED ISSUE #140: Service provider pretty name is shared
See from my talk https://youtu.be/Uv0ivz5xej4?t=1213, look at the columns. Something is wrong:

Bildschirm­foto 2022-11-21 um 19 16 30

### PR DESCRIPTION
**Pull request by pkbhowmick**
Signed-off-by: Pulak Kanti Bhowmick <pulak@appscode.com>

Fix #140 

As konnector is running APIServiceBinding controller for each provider cluster, so each controller should skip others provider's APIServiceBinding object.

### CODE REVIEW

File: pkg/konnector/controllers/cluster/servicebinding/servicebinding_reconcile.go [hunk 1]
```diff
@@ -46,7 +50,38 @@ type reconciler struct {
...
+	if !found {
+		logger.Error(err, "kubeconfig in secret does not have a current context", "namespace", ref.Namespace, "name", ref.Name)
+		return nil // nothing we can do here. The APIServiceBinding Controller will set a condition
+	}
+
+	// As konnector is running APIServiceBinding controller for each provider cluster,
+	// so each controller should skip others provider's APIServiceBinding object
+	if kubeContext.Namespace != r.providerNamespace {
+		return nil
+	}
```
Comment (sttts) [inline pkg/konnector/controllers/cluster/servicebinding/servicebinding_reconcile.go:84]: wondering, is this even enough? Couldn't there be multiple connected provider clusters with the same namespace name? I would rather store the APIServiceBinding name here in the controller instance and skip other service bindings.

Comment (pkbhowmick) [reply pkg/konnector/controllers/cluster/servicebinding/servicebinding_reconcile.go:84]: yes, there can be multiple clusters with the same namespace name. so it is better to keep the service binding name in controller instance. Updating...
Thanks!

Comment (pkbhowmick) [reply pkg/konnector/controllers/cluster/servicebinding/servicebinding_reconcile.go:55]: Updated with service binding name.

File: pkg/konnector/konnector_reconcile.go [hunk 2]
```diff
@@ -120,6 +120,7 @@ func (r *reconciler) reconcile(ctx context.Context, binding *kubebindv1alpha1.AP
 	ctrl, err := r.newClusterController(
 		binding.Spec.KubeconfigSecretRef.Namespace+"/"+binding.Spec.KubeconfigSecretRef.Name,
 		providerNamespace,
+		binding.Name,
```
Comment (sttts) [inline pkg/konnector/konnector_reconcile.go:123]: it's a little bit more complicated. Compare the `ctrlContext` logic above. There The controller context contains a list of bindings it is started for. So it is also responsible for all its bindings, i.e. not only `binding.Name` here.

File: pkg/konnector/konnector_reconcile.go [hunk 3]
```diff
@@ -120,6 +120,9 @@ func (r *reconciler) reconcile(ctx context.Context, binding *kubebindv1alpha1.AP
 	ctrl, err := r.newClusterController(
 		binding.Spec.KubeconfigSecretRef.Namespace+"/"+binding.Spec.KubeconfigSecretRef.Name,
 		providerNamespace,
+		func(svcBinding *kubebindv1alpha1.APIServiceBinding) bool {
+			return svcBinding.Name == binding.Name
```
Comment (sttts) [inline pkg/konnector/konnector_reconcile.go:124]: nearly. You have to check that `svcBinding.Name` is in `ctrlContext.serviceBindings`, while taking the `r.lock`. Maybe it's a good idea to turn the lock into a RWMutex.

File: pkg/konnector/konnector_reconcile.go [hunk 4]
```diff
@@ -120,6 +120,16 @@ func (r *reconciler) reconcile(ctx context.Context, binding *kubebindv1alpha1.AP
 	ctrl, err := r.newClusterController(
 		binding.Spec.KubeconfigSecretRef.Namespace+"/"+binding.Spec.KubeconfigSecretRef.Name,
 		providerNamespace,
+		func(svcBinding *kubebindv1alpha1.APIServiceBinding) bool {
+			r.lock.RLock()
+			defer r.lock.RUnlock()
+			// for the very first time, ctrlContext is nil
+			// and need to check for this as well
```
Comment (sttts) [inline pkg/konnector/konnector_reconcile.go:127]: don't think this is necessary. The controller starts in line 146 after the `r.controllers` change.

File: pkg/konnector/konnector_reconcile.go [hunk 5]
```diff
@@ -120,6 +120,16 @@ func (r *reconciler) reconcile(ctx context.Context, binding *kubebindv1alpha1.AP
...
+			if binding.Name == svcBinding.Name {
+				return true
+			}
+			return ctrlContext != nil && ctrlContext.serviceBindings.Has(svcBinding.Name)
```
Comment (sttts) [inline pkg/konnector/konnector_reconcile.go:131]: This won't work here. You have to move the whole block of line 140-145 above this closure.

**Review by sttts — APPROVED** [review]

**Pull request merged**